cam6_3_066: cleaning up clubb bugs, misc clubb issues

adamrher opened the pull request:
Fixes #336 
Fixes #496 
Fixes #497
Fixes #536 
Fixes #582 
Fixes #604 
Fixes #605

This PR is an continuation of (closed) PR https://github.com/ESCOMP/CAM/pull/352

adamrher commented:
@Katetc @JulioTBacmeister @bstephens82 @swrneale I'd say about half of the CLUBB outfld vars are incorrectly assigned to "ilev" when they are technically on the "lev". This is because in clubb, the midpoint grid contains pverp entries, just like the interface grid, because there is a ghost point in the subsurface. 

Does everyone agree with putting these outfld vars on the correct grid? Or leave it as it is?

JulioTBacmeister commented:
I think the CLUBB output should be on the correct grid.

Katetc commented:
Does that mean we won't output the ghost point? That's probably fine. I can't remember if it was ever important. Be careful with the array truncation code, that's an easy thing to get weird seg fault errors in.

adamrher commented:
> Does that mean we won't output the ghost point?

Yes. Which is fine IMO too. The clubb documentation says that after the state is advanced in clubb, it just copies the pver values into pverp (the ghost point). So there's no additional information.

I'll plan on fixing the outfld's then...

bstephens82 commented:
Sounds good to me.

bstephens82 commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -2614,6 +2583,7 @@ subroutine clubb_tend_cam( &
      do i=1,ncol
        inv_exner_clubb(i,k) = 1._r8/((state1%pmid(i,k)/p0_clubb)**(rairv(i,k,lchnk)/cpairv(i,k,lchnk)))
      enddo
+     inv_exner_clubb_surf(i) = 1._r8/((state1%pint(i,pverp)/p0_clubb)**(rair/cpair))
```
This statement contains an `i` but is outside of the `i` loop...is that intentional?

adamrher replied:
good catch! yep this is a mistake; wasn't looking closely at the loop nesting

bstephens82 replied:
Resolved.

bstephens82 commented on src/physics/cam/check_energy.F90:
```diff
@@ -66,6 +66,7 @@ module check_energy
 
   integer  :: teout_idx  = 0       ! teout index in physics buffer
   integer  :: dtcore_idx = 0       ! dtcore index in physics buffer
+  integer  :: dqcore_idx = 0
```
Should this be commented?

adamrher replied:
fixed

bstephens82 replied:
Resolved.

bstephens82 commented on src/physics/cam/physpkg.F90:
```diff
@@ -88,6 +88,8 @@ module physpkg
   integer ::  ducore_idx         = 0     ! ducore index in physics buffer
   integer ::  dvcore_idx         = 0     ! dvcore index in physics buffer
+  integer ::  dtcore_idx         = 0
+  integer ::  dqcore_idx         = 0
```
Should these two lines be commented?

adamrher replied:
done

bstephens82 replied:
Resolved.

bstephens82 commented on src/physics/cam_dev/physpkg.F90:
```diff
@@ -86,6 +86,7 @@ module physpkg
   integer ::  dvcore_idx         = 0     ! dvcore index in physics buffer
   integer ::  dtcore_idx         = 0
+  integer ::  dqcore_idx         = 0
   integer ::  cmfmczm_idx        = 0     ! Zhang-McFarlane convective mass fluxes
```
Should this be commented?

bstephens82 replied:
Resolved.

cacraigucar reviewed (commented):
First pass review

adamrher commented on src/physics/cam/cam_diagnostics.F90:
```diff
@@ -1344,8 +1350,32 @@ subroutine diag_phys_writeout_moist(state, pbuf, p_surf_t)
+    call outfld ('IVT     ',ftem, pcols   ,lchnk     )
+
+    ! output uq*dp/g
+    ftem(:ncol,1) =  ftem4(:ncol,1)
+    call outfld ('uIVT     ',ftem, pcols   ,lchnk     )
```
no reason. I removed this redundancy and just put ftem4/5 in the outfld calls directly.

adamrher commented on src/physics/cam/micro_pumas_cam.F90:
```diff
@@ -48,11 +48,13 @@ module micro_pumas_cam
 
 integer :: micro_mg_version     = 1      ! Version number for MG.
 integer :: micro_mg_sub_version = 0      ! Second part of version number.
+integer :: micro_mg_num_steps   = 1      ! Number of substepping iterations done by MG (versions 1.5 and later).
```
thanks! I actually pulled this out of the branch because it relates to outputting the time-steps to the log. But I made a new branch with just the time-step output code, and added your mods to micro_pumas_cam.

adamrher commented on src/physics/cam/check_energy.F90:
```diff
@@ -139,11 +140,13 @@ subroutine check_energy_register()
 
     call pbuf_add_field('TEOUT', 'global',dtype_r8 , (/pcols,dyn_time_lvls/),      teout_idx)
     call pbuf_add_field('DTCORE','global',dtype_r8,  (/pcols,pver,dyn_time_lvls/),dtcore_idx)
+    call pbuf_add_field('DQCORE','global',dtype_r8,  (/pcols,pver,dyn_time_lvls/),dqcore_idx)
```
done

adamrher commented on src/physics/cam/check_energy.F90:
```diff
@@ -199,6 +202,7 @@ subroutine check_energy_init()
     call addfld('TEFIX',  horiz_only,  'A', 'J/m2', 'Total energy after fixer')
     call addfld('EFIX',   horiz_only,  'A', 'W/m2', 'Effective sensible heat flux due to energy fixer')
     call addfld('DTCORE', (/ 'lev' /), 'A', 'K/s' , 'T tendency due to dynamical core')
+    call addfld('DQCORE', (/ 'lev' /), 'A', 'kg/kg/s' , 'Q tendency due to dynamical core')
```
changed it to "Water vapor tendency ..."

adamrher commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -2953,8 +2933,10 @@ subroutine clubb_tend_cam( &
         rtpthlp_in(i,k) = rtpthlp(i,pverp-k+1)
-        rcm_inout(i,k)  = rcm(i,pverp-k+1)
         cloud_frac_inout(i,k) = cloud_frac(i,pverp-k+1)
+        if (k.gt.1) then
```
done (ive been reminded of this before ... hopefully it will take this time)

adamrher commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -4137,13 +4115,15 @@ subroutine clubb_tend_cam( &
-         thv(i,k) = th(i,k)*(1.0_r8+zvir*state1%q(i,k,ixq))
+         !th(i,k) = state1%t(i,k)*state1%exner(i,k)
+         !thv(i,k) = th(i,k)*(1.0_r8+zvir*state1%q(i,k,ixq))
+         ! REMOVE AFTER CODE REVIEW - state%exner is not a proper exner
+         !                            thv should have condensate loading to be consistent with earlier def's
```
done

adamrher commented:
@cacraigucar this PR is almost ready for prime time. Three outstanding issues:

1. @Katetc are there are low hanging clubb issues that we should add to this PR?
2. Do I need to merge up to the head? My branch is currently based on cam6_3_060.
3. Currently doing a quick science eval of these changes, which should be completed by tomorrow.

cacraigucar commented:
> 2. Do I need to merge up to the head? My branch is currently based on cam6_3_060.

Yes, this will need to be merged to the head of cam_development before we run the final regression testing (once the reviews are complete and we are ready to begin the tagging process).  The code is fairly close to the head, so not critical for the review stage, especially since the tags since then have been pretty orthogonal.

adamrher commented:
@JulioTBacmeister
@swrneale 
@andrewgettelman 
@bstephens82

Impact of this PR is [here](https://webext.cgd.ucar.edu/F2000climo/clubb.kitchensink_F2000dev_ne30pg3_ne30pg3_mg17_L58dev_nuopc_3600pes_220623_Nx2yrs/atm/clubb.kitchensink_F2000dev_ne30pg3_ne30pg3_mg17_L58dev_nuopc_3600pes_220623_Nx2yrs.2_8-cam6_3_060_F2000dev_ne30pg3_ne30pg3_mg17_L58dev_nuopc_3600pes_220623_Nx2yrs.2_8/) --8 years, L58, F2000dev. Seems to be a bit more convection in W.Pac, and an increase in ice water path. Nothing alarming.

swrneale commented:
West Pacific and Indian Monsoon are quite a bit wetter, which makes biases a bit worse. But in the far East Pacific it's slightly better. Mixed bag I guess. Do you have expectations of how stable/unstable the state is it on exit from CLUBB when you change the position of the PDF closure?

adamrher commented:
wait a second I see something wrong - [PBLH](https://webext.cgd.ucar.edu/F2000climo/clubb.kitchensink_F2000dev_ne30pg3_ne30pg3_mg17_L58dev_nuopc_3600pes_220623_Nx2yrs/atm/clubb.kitchensink_F2000dev_ne30pg3_ne30pg3_mg17_L58dev_nuopc_3600pes_220623_Nx2yrs.2_8-cam6_3_060_F2000dev_ne30pg3_ne30pg3_mg17_L58dev_nuopc_3600pes_220623_Nx2yrs.2_8/set5_6/set5_ANN_PBLH_c.png).

I had previously tested the impact of the pdf closure position and not found any appreciably change. Rather it seems like I messed up the PBLH calculation in clubb_intr.

swrneale commented:
Is it using the Holslag Boville PBLH calculation still? Does the PDF closure come at a difference place relative to this calculation now?

adamrher commented:
Yes to Holstag Boville. No the pdf closure is not in a different place from the perspective of the PBLH call.

It's got to be one of two things. I changed the definition of theta to use the true exner function (not state%exner), and then in the theta_v calculation I added condensate loading. Doing two simulations with each of these reverted back right now. Note these theta / theta_v calculations are only used by the PBLH calculation.

adamrher commented:
I fixed this bug I introduced into the PBLH calculation. The problem was that I naively just replaced state%exner with the exner defined in clubb_intr. But state%exner is cryptically named -- it is actually 1/exner. This has tripped up many of us before, and so should probably be fixed.

So this is fixed, and I'm re-running 10 years to evaluate the impact of this PR.

adamrher commented:
Ten year runs of L58, cam_dev, [showing impact of this PR](https://webext.cgd.ucar.edu/F2000climo/clubb.kitchensink_F2000dev_ne30pg3_ne30pg3_mg17_L58dev_nuopc_3600pes_220625_fixexner2_Nx2yrs/atm/clubb.kitchensink_F2000dev_ne30pg3_ne30pg3_mg17_L58dev_nuopc_3600pes_220625_fixexner2_Nx2yrs.1_9-cam6_3_060_F2000dev_ne30pg3_ne30pg3_mg17_L58dev_nuopc_3600pes_220623_Nx2yrs.1_9/). @swrneale there are no precip changes in tropical pacific or indian monsoon region anymore. The only changes I see are TS, primarily over land. Interestingly TREFHT has much smaller changes. This resembles what we found when we tested the impact of [scaling sensible heat fluxes in isolation](https://webext.cgd.ucar.edu/F2000climo/cam6_2_026.pel_F2000climo_ne30pg3_ne30pg3_mg17_1800pes_210924_ediags_scaleshf_Nx5yrs.arh/atm/cam6_2_026.pel_F2000climo_ne30pg3_ne30pg3_mg17_1800pes_210924_ediags_scaleshf_Nx5yrs.2_20-cam6_2_026.pel_F2000climo_ne30pg3_ne30pg3_mg17_1800pes_210924_ediags_Nx5yrs.2_20/set5_6/set5_ANN_TS_c.png). Note these old expts are L32 runs, whereas these new expts are L58 cam_dev. The largest TS changes are over high elevation regions, the Himalayas and the big ice sheets. This is where we expect the largest changes since dividing by exner is only going to impact regions where PS is very different from P0=1000 hPa.

Recall also that the scaling of the sensible heat fluxes drastically reduces CLUBB's energy leak, so this is a substantial enhancement to the model. See Figure 12 of @PeterHjortLauritzen's very short little GRL paper [here](https://www.cgd.ucar.edu/cms/pel/papers/LetAl2022JAMES.pdf).

cacraigucar commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -2921,6 +2899,9 @@ subroutine clubb_tend_cam( &
     do i=1,ncol                                                                  
       wpthlp_sfc(i) = cam_in%shf(i)/(cpair*rho_ds_zm(i,1))       ! Sensible heat flux
+      !  Convert to flux of potential temperature
+      !  (sec 4.3.2 in Lauritzen et al. 2022, JAMES)
```
Do you have a DOI for this reference?  We've started trying to put those in as well.

adamrher replied:
it's still in review, so no doi yet, unfortunately. and section numbering might change. maybe we should just remove line 2903 for now?

cacraigucar commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -3969,27 +3961,14 @@ subroutine clubb_tend_cam( &
    !  density
-   rho(:ncol,1:pver) = state1%pmid(:ncol,1:pver)/(rairv(:ncol,1:pver,lchnk)*state1%t(:ncol,1:pver))
-   rho(:ncol,pverp)  = state1%ps(:ncol)/(rairv(:ncol,pver,lchnk)*state1%t(:ncol,pver))
+   ! REMOVE AFTER CODE REVIEW - this rho is for midpoints except the lowest level is the surface
+   !                          - BUT this is supposed to be rho at the interface levels
+   !                          - replace with rho at mid-points using mass def, to be consistent w/ earlier def's
```
Just a reminder to remove this after all the reviewers have seen it.  I assume you will be removing the commented out lines as well.

adamrher replied:
ok. will remove when everyone has done a first pass.

cacraigucar commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -4172,64 +4151,59 @@ subroutine clubb_tend_cam( &
 
-     temp2dp(:ncol,:) = rcm(:ncol,:) * 1000._r8
-     call outfld( 'RCM_CLUBB',        temp2dp,                 pcols, lchnk )
+     temp2d(:ncol,:) = thlm(:ncol,1:pver)
+     call outfld( 'THLM_CLUBB',       temp2d,                 pcols, lchnk )
```
Since the arithmetic was removed, you can remove the copy operation and just put the thlm in the outfld call directly.  This comment applies to the rcm_in_layer below as well.

adamrher replied:
ok. same goes for existing entires before and after this line. Removed all unnecessary copy operations nearby.

cacraigucar reviewed (commented):
A few more minor points.  I think this will be ready for full review once you remove it from draft status.

cacraigucar reviewed (commented):
One more general comment.  I believe that the description needs one issue number following a keyword like "fixes" to get the automation to close the issue.  @nusbaume is that correct, or can you have a list of issues?

nusbaume commented:
@cacraigucar @adamrher Sadly you do need to have the "fixes" statement repeated for each issue, as well as just the issue number alone with the "#" sign (not the full URL).  

I went ahead and updated the main PR description to match these requirements, so hopefully that was ok!

nusbaume reviewed (changes requested):
Looks good, but I do have some minor code clean-up requests.

adamrher commented on src/physics/cam/cam_diagnostics.F90:
```diff
@@ -1344,8 +1350,30 @@ subroutine diag_phys_writeout_moist(state, pbuf, p_surf_t)
+    ! Integrated vapor transport calculation
+    !
+    !compute uq*dp/g and vq*dp/g
+    ftem4(:ncol,:) = state%q(:ncol,:,1) * state%u(:ncol,:) *state%pdel(:ncol,:) * rga
+    ftem5(:ncol,:) = state%q(:ncol,:,1) * state%v(:ncol,:) *state%pdel(:ncol,:) * rga
```
OK. This does seem to be pervasive throughout this subroutine. I added in ixq / get_index_pbuf locally to the subroutine, and changed all occurrences of ```state%q(:ncol,:,1)``` to ```state%q(:ncol,:,ixq)```. I didn't try to define ixq at the module level because ixq is a intent(in) argument to one of the other subroutines in the module, which would create duplicate definitions.

nusbaume replied:
No worries, I actually only expected that you would modify the new code, so way to go above and beyond!

adamrher commented on src/physics/cam/cam_diagnostics.F90:
```diff
@@ -1290,6 +1294,8 @@ subroutine diag_phys_writeout_moist(state, pbuf, p_surf_t)
     real(r8) :: ftem(pcols,pver) ! temporary workspace
     real(r8) :: ftem1(pcols,pver) ! another temporary workspace
     real(r8) :: ftem2(pcols,pver) ! another temporary workspace
+    real(r8) :: ftem4(pcols,pver) ! another temporary workspace
+    real(r8) :: ftem5(pcols,pver) ! another temporary workspace
```
replaced these with existing ftem1 and ftem2 arrays.

adamrher commented on src/physics/cam_dev/physpkg.F90:
```diff
@@ -2597,9 +2604,11 @@ subroutine tphysbc (ztodt, state,  &
     ! T, U, V tendency due to dynamics
     if ( nstep > dyn_time_lvls-1 ) then
        dtcore(:ncol,:pver) = (state%t(:ncol,:pver) - dtcore(:ncol,:pver))/ztodt
+       dqcore(:ncol,:pver) = (state%q(:ncol,:pver,1) - dqcore(:ncol,:pver))/ztodt
```
done

adamrher commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -2755,12 +2721,21 @@ subroutine clubb_tend_cam( &
+                                   /(1._r8 + state1%q(i,pver-k+1,ixq)*rh2o/rairv(i,pver-k+1,lchnk))
+        exner_ds_zt(i,k+1)     = (p_ds_zt(i,k+1)/p0_clubb)**(rairv(i,pver-k+1,lchnk)/cpairv(i,pver-k+1,lchnk))
+        thv_ds_zt(i,k+1)       = state1%t(i,pver-k+1)/exner_ds_zt(i,k+1)
+        rho_ds_zt(i,k+1)       = (1._r8/gravit)*(state1%pdel(i,pver-k+1) &
```
done

adamrher commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -3969,27 +3961,14 @@ subroutine clubb_tend_cam( &
+   !                          - replace with rho at mid-points using mass def, to be consistent w/ earlier def's
+   !rho(:ncol,1:pver) = state1%pmid(:ncol,1:pver)/(rairv(:ncol,1:pver,lchnk)*state1%t(:ncol,1:pver))
+   !rho(:ncol,pverp)  = state1%ps(:ncol)/(rairv(:ncol,pver,lchnk)*state1%t(:ncol,pver))
+   rho(1:ncol,1:pver) = (1._r8/gravit)*state1%pdel(1:ncol,1:pver)/(state1%zi(1:ncol,1:pver)-state1%zi(1:ncol,2:pverp))
```
done

nusbaume reviewed (approved):
Looks great now, thanks!

cacraigucar commented on src/physics/cam/physpkg.F90:
```diff
@@ -1912,6 +1916,7 @@ subroutine tphysac (ztodt,   cam_in,  &
     ! store T, U, and V in buffer for use in computing dynamics T-tendency in next timestep
     do k = 1,pver
        dtcore(:ncol,k) = state%t(:ncol,k)
+       dqcore(:ncol,k) = state%q(:ncol,k,ixq)
```
ixq is not in scope (defined) in tphysac.  I'm working on a mod to fix this.  The cam_dev version already had the ixq defined in tphysac, which is why this was missed.

adamrher replied:
ah right, only in cam_dev/physpkg.F90 is ixq defined at the tphysac level.

cacraigucar commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -4172,66 +4149,48 @@ subroutine clubb_tend_cam( &
-   call outfld( 'RELVAR',           relvar,                  pcols, lchnk )
-   call outfld( 'RHO_CLUBB',        rho,                   pcols, lchnk )
+   call outfld( 'RELVAR',           relvar,                pcols, lchnk )
+   call outfld( 'RHO_CLUBB',        rho(1:ncol,1:pver),    pcols, lchnk )
```
At least two regression tests are erroring out at this location:

[11] Runtime Error: *** Arithmetic exception: Floating invalid operati[11] on - aborting
[11] /home/cacraig/cam_clubb_kitchensink/src/control/cam_history_buffers.F90, line 104: Error occurred[11]  in CAM_HISTORY_BUFFERS:HBUF_ACCUM_ADD
[11] /home/cacraig/cam_clubb_kitchensink/src/control/cam_history.F90, line 3429: Called by CAM_HISTORY:OUTFLD
[11] /home/cacraig/cam_clubb_kitchensink/src/physics/cam/clubb_intr.F90, line 4153: C[11] alled by CLUBB_INTR:CLUBB_TEND_CAM

A run can be seen at: 
/scratch/cluster/cacraig/aux_cam_nag_20220705130436/ERC_D_Ln9_Vnuopc.f10_f10_mg37.QPWmaC6.izumi_nag.cam-outfrq3s.GC.aux_cam_nag_20220705130436/run

@adamrher  - Can you take a look at this?  I can check into this tomorrow if you don't solve this.  I'm guessing a mismatch in dimensions possibly.

adamrher replied:
can you provide a create test command to reproduce this error?

cacraigucar replied:
The create_newcase used was:

 /home/cacraig/cam_clubb_kitchensink/cime/CIME/scripts/create_newcase.py --case /scratch/cluster/cacraig/aux_cam_nag_20220705130436/ERC_D_Ln9_Vnuopc.f10_f10_mg37.QPWmaC6.izumi_nag.cam-outfrq3s.GC.aux_cam_nag_20220705130436 --res f10_f10_mg37 --compset QPWmaC6 --test --machine izumi --compiler nag --output-root /scratch/cluster/cacraig/aux_cam_nag_20220705130436 --workflow default --srcroot /home/cacraig/cam_clubb_kitchensink --driver nuopc --user-mods-dir /home/cacraig/cam_clubb_kitchensink/cime_config/testdefs/testmods_dirs/cam/outfrq3s --walltime 00:30:00

I know you need to remove the --test and --workflow default since you won't be in the testing suite.

cacraigucar replied:
> isn't there a much easier way to run this test? Say:
> 
> ./create_test ERC_D_Ln9_Vnuopc.f10_f10_mg37.QPWmaC6.izumi_nag.cam-outfrq3s
> 
> ??

Yes, you can use that command as well

adamrher replied:
It looks to be throwing a fit over all my new array subsetting in the outfld calls. E.g., this gives an error:
```
   call outfld( 'RHO_CLUBB',        rho(1:ncol,1:pver),  pcols, lchnk )
```
This does not:
```
   call outfld( 'RHO_CLUBB',        rho,                   pcols, lchnk )
```
I introduced this (1:ncol,1:pver) subsetting of arrays for about 10 different outfld calls in this PR. These arrays are of size (pcol,pverp). The problem is that clubb's grid for the mid-point levels has pverp entries, whereas CAM's mid-point grid only has pver entries, because clubb plops an additional vertical level in the subsurface. So my goal is to omit the pverp / subsurface point, and only outfld the 1:pver entries of the arrays, since they should be on CAM's  ```lev``` grid. 

@cacraigucar can you take a look at this tmrw?

gold2718 replied:
> It looks to be throwing a fit over all my new array subsetting in the outfld calls. E.g., this gives an error:
> 
> ```
>    call outfld( 'RHO_CLUBB',        rho(1:ncol,1:pver),  pcols, lchnk )
> ```

Makes sense, try:
```
   call outfld( 'RHO_CLUBB',        rho(1:ncol,1:pver),  ncol, lchnk )
```
Note the change in the third argument (which is the size of the first dimension of the second argument).
